File: android/app/src/main/java/com/mirrorfly_rn/ActivityModule.java
// Adapted from
// https://github.com/gijoehosaphat/react-native-keep-screen-on

package com.mirrorfly_rn;

import android.app.Activity;
import android.app.KeyguardManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.util.Log;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.List;

public class ActivityModule extends ReactContextBaseJavaModule {

    public ReactApplicationContext reactContext;
    public static ActivityModule instance = null;
    public static final String CALL_SCREEN_ACTIVITY_STATE_CHANGE = "CALL_SCREEN_ACTIVITY_STATE_CHANGE";
    private static DeviceEventManagerModule.RCTDeviceEventEmitter eventEmitter = null;

    public ActivityModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public void initialize() {
        super.initialize();
        eventEmitter = getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class);
    }

    @Override
    public String getName() {
        return "ActivityModule";
    }

    public static synchronized void callScreenStateChanged(String state) {
        Log.d("TAG", "callScreenStateChanged: emitting event callScreenStateChanged");
        eventEmitter.emit(CALL_SCREEN_ACTIVITY_STATE_CHANGE, state);
    }

    @ReactMethod
    public void closeActivity() {
        ActivityManager.getInstance().finishActivity(CallScreenActivity.class);
    }

    @ReactMethod
    public void updateCallConnectedStatus(Boolean isConnected) {
        CallScreenActivity.isCallConnected = isConnected;
    }

    @ReactMethod
    public void getPackageName(Promise promise) {
        final Activity activity = getReactApplicationContext().getCurrentActivity();
        if (activity != null) {
            try {
                String packageName = activity.getPackageName();
                promise.resolve(packageName);
            } catch (Exception e) {
                promise.resolve("undefined");
            }
        } else {
            promise.resolve("undefined");
        }
    }

    @ReactMethod
    public void getInstalledPackages(String targetPackage, Promise promise) {
        Boolean isPackageExists = isPackageExists(targetPackage);
        promise.resolve(isPackageExists);
    }

    public boolean isPackageExists(String targetPackage) {
        PackageManager pm = getReactApplicationContext().getPackageManager();
        List<ApplicationInfo> packages = pm.getInstalledApplications(0);
        for (ApplicationInfo packageInfo : packages) {
            if (packageInfo.packageName.equals(targetPackage)) {
                return true;
            }
        }
        return false;
    }

    @ReactMethod
    public void getActivity(Promise promise) {
        final Activity activity = this.reactContext.getCurrentActivity();
        if (activity != null) {
            try {
                String classname = activity.getComponentName().getClassName();
                promise.resolve(classname);
            } catch (Exception e) {
                promise.resolve("undefined");
            }
        } else {
            promise.resolve("undefined");
        }
    }

    @ReactMethod
    public void getCallActivity(Promise promise) {
        try {
            final String activity = CallScreenActivity.class.getName();
            promise.resolve(activity);
        } catch (Exception e) {
            promise.resolve("undefined");
        }
    }

    @ReactMethod
    public void getMainActivity(Promise promise) {
        try {
            final String activity = MainActivity.class.getName();
            promise.resolve(activity);
        } catch (Exception e) {
            promise.resolve("undefined");
        }
    }

    @ReactMethod
    public void isLocked(Promise promise) {
        try {
            boolean isLocked = isLockScreenDisabled();
            promise.resolve(isLocked);
        } catch (Exception e) {
            promise.reject(e);
        }
    }

    private boolean isLockScreenDisabled() {
        KeyguardManager km = (KeyguardManager) this.reactContext.getSystemService(Context.KEYGUARD_SERVICE);
        if (km.isKeyguardLocked())
            return true;
        else
            return false;
    }

    @ReactMethod
    public void openActivity() {
        Intent intent = new Intent(getReactApplicationContext(), CallScreenActivity.class);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getReactApplicationContext().startActivity(intent);
    }
}

File: android/app/src/main/java/com/mirrorfly_rn/ActivityModule.kt
// Adapted from
// https://github.com/gijoehosaphat/react-native-keep-screen-on
package com.mirrorfly_rn

import android.app.KeyguardManager
import android.content.Context
import android.content.Intent
import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.mirrorfly_rn.CallScreenActivity

class ActivityModule(var reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(
    reactContext
) {
    override fun initialize() {
        super.initialize()
        eventEmitter = reactApplicationContext.getJSModule(
            DeviceEventManagerModule.RCTDeviceEventEmitter::class.java
        )
    }

    override fun getName(): String {
        return "ActivityModule"
    }

    @ReactMethod
    fun closeActivity() {
        ActivityManager.instance?.finishActivity(CallScreenActivity::class.java)
    }

    @ReactMethod
    fun updateCallConnectedStatus(isConnected: Boolean) {
        CallScreenActivity.isCallConnected = isConnected
    }

    @ReactMethod
    fun getPackageName(promise: Promise) {
        val activity = reactApplicationContext.currentActivity
        if (activity != null) {
            try {
                val packageName = activity.packageName
                promise.resolve(packageName)
            } catch (e: Exception) {
                promise.resolve("undefined")
            }
        } else {
            promise.resolve("undefined")
        }
    }

    @ReactMethod
    fun getInstalledPackages(targetPackage: String, promise: Promise) {
        val isPackageExists = isPackageExists(targetPackage)
        promise.resolve(isPackageExists)
    }

    fun isPackageExists(targetPackage: String): Boolean {
        val pm = reactApplicationContext.packageManager
        val packages = pm.getInstalledApplications(0)
        for (packageInfo in packages) {
            if (packageInfo.packageName == targetPackage) {
                return true
            }
        }
        return false
    }

    @ReactMethod
    fun getActivity(promise: Promise) {
        val activity = reactContext.currentActivity
        if (activity != null) {
            try {
                val classname = activity.componentName.className
                promise.resolve(classname)
            } catch (e: Exception) {
                promise.resolve("undefined")
            }
        } else {
            promise.resolve("undefined")
        }
    }

    @ReactMethod
    fun getCallActivity(promise: Promise) {
        try {
            val activity = CallScreenActivity::class.java.name
            promise.resolve(activity)
        } catch (e: Exception) {
            promise.resolve("undefined")
        }
    }

    @ReactMethod
    fun getMainActivity(promise: Promise) {
        try {
            val activity = MainActivity::class.java.name
            promise.resolve(activity)
        } catch (e: Exception) {
            promise.resolve("undefined")
        }
    }

    @ReactMethod
    fun isLocked(promise: Promise) {
        try {
            val isLocked = isLockScreenDisabled
            promise.resolve(isLocked)
        } catch (e: Exception) {
            promise.reject(e)
        }
    }

    private val isLockScreenDisabled: Boolean
        get() {
            val km =
                reactContext.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
            // boolean isLocked = km.inKeyguardRestrictedInputMode();
            // if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            // Log.d("isLocked", "isLockScreenDisabled: isLocked " + isLocked + " ,, " +
            // km.isKeyguardSecure() +" , "+ km.isDeviceSecure());
            // }
            return km.isKeyguardLocked
        }

    @ReactMethod
    fun addListener(eventName: String?) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    fun removeListeners(count: Int?) {
        // Keep: Required for RN built in Event Emitter Calls.
    }

    @ReactMethod
    fun openActivity() {
        val intent = Intent(
            reactApplicationContext,
            CallScreenActivity::class.java
        )
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        reactApplicationContext.startActivity(intent)
    }

    companion object {
        const val CALL_SCREEN_ACTIVITY_STATE_CHANGE: String = "CALL_SCREEN_ACTIVITY_STATE_CHANGE"
        private var eventEmitter: DeviceEventManagerModule.RCTDeviceEventEmitter? = null

        @Synchronized
        fun callScreenStateChanged(state: String?) {
            Log.d("TAG", "callScreenStateChanged: emitting event callScreenStateChanged")
            eventEmitter!!.emit(CALL_SCREEN_ACTIVITY_STATE_CHANGE, state)
        }
    }
}